Python 3 style conversion

jameshilliard opened the pull request:
This introduces the [future](https://python-future.org/) library which backports many python3 features. I'm using this to force as much python3 style as possible without breaking python2 compatibility.

For larger python projects like this(which has especially painful bytes/unicode issues) a flag day conversion is not really feasible without severe disruption so the strategy I'm using here is to incrementally eliminate python2 only features/styles and replace them with the back-ported python3 versions from the future library. This allows the test suite to be run using python2 after porting individual files.

Once the project is converted to the backported python3 style it should be feasible to fix any remaining compatibility issues and run the test suite on both python2 and python3. After the tests function on both python versions for long enough to weed out regressions we can then drop python2 support.

undeath commented:
The changes look good, except for the ones in `jmbase/jmbase/commands.py`.

Changing all `str` occurrences in the `Command.arguments` and `Command.response` containers to `bytes` seems wrong to me. The twisted documentation doesn't explicitly state what data type those "names" should be, but "name", to me, implies a `str`, not `bytes`. On what basis did you change that?

The documentation for `Command.errors` explicitly mentions expecting a `str` value, not `bytes`.

Moving from twisted's `String` to `Unicode` makes sense to me. This would, as far as I understand, cause errors that would appear in Python 3 to also appear in Python 2. On the other hand, we would of course not want any errors to occur there at all.

How did you decide what files to change? The subset of files and changes made seem a little random. For example there is only one file where you changed `xrange`. Wouldn't a more structured approach be more helpful?

undeath commented on jmbase/jmbase/commands.py:
```diff
@@ -157,47 +159,47 @@ class JMRequestMsgSig(JMCommand):
     """Return the entire contents of the
     orderbook to TAKER, as a json-ified dict;
-    note uses BigString because can be very large
+    note uses BigUnicode() because can be very large
```
the data type is still called BigString

jameshilliard replied:
Fixed, I had forgotten to fix that from the find and replace.

AdamISZ commented:
As discussed on IRC, I like the incremental approach. As for the files here, I guess you're addressing some but not all of the incompatibilities in specifically `jmbase` and `jmdaemon`, is that right?

Re: BigString, it seems that in the (nearly 2 years) time since I wrote this, the documentation is updated and it appears that AMP v2 supports arbitrary sized strings intrinsically anyway now. See [here](https://amp-protocol.net/dontpanic.html?highlight=bigstring).

AdamISZ commented:
To echo what I've probably written elsewhere, when it comes to jmbitcoin at least, it probably makes sense to fold in the upgrade to Python3 compatibility with switching the binding to libsecp256k1 if possible, see coincurve (although such of thing cannot, of course, be done lightly). Since secp256k1-py is no longer maintained (which is more relevant than in some other cases, given the status of libsecp256k1). Segwit support was somewhat tacked-on and the weaknesses of the library (which started out as a port of Buterin's pybitcointools) are certainly showing now. It could do with a major rewrite in parts.

There is also an argument for a full-on rewrite of jmclient and jmdaemon using asyncio ... I don't really have the knowledge to assess this idea, but clearly it sounds like a lot of work.

This might also be a time to overhaul tests, which I made initially in a pretty messy way, tried to improve at one time, but then they suffered from pretty serious code-rot over the years.

Apologies for the slight braindump in here, but it could be worth discussing at that higher level what any devs want to do in this area, in case we accidentally end up putting efforts into porting code that isn't going to end up being used.

jameshilliard commented:
> On what basis did you change that?

Mostly since it's a command for a wire protocol which typically implies bytes for native format(some libraries have automatic conversions but bytes is usually clearer I think). It also uses bytes in the [example](https://github.com/twisted/twisted/blob/6ac66416c0238f403a8dc1d42924fb3ba2a2a686/docs/core/examples/ampserver.py#L12-L13).

> The documentation for Command.errors explicitly mentions expecting a str value, not bytes.

I'm using the `unicode_literals` import which should ensure that there's no python2 `str` type available that can be used(since python2's `str` doesn't exist in python3 anymore), only `unicode` or `bytes`.

> How did you decide what files to change?

So I was trying to run some tests against python3 and fix issues from that but was running into too many encoding errors, mostly in `jmbitcoin`. I tried to incorporate some of those fixes I found for python3 errors I found in `jmbase` and `jmdaemon` into this but mostly I just tried adding those imports to a module and then fix encoding issues. The main issues is that things are just too broken in python3 right now to properly test there so I'm trying to reduce the errors while mostly working in a python2 environment.

> For example there is only one file where you changed xrange. Wouldn't a more structured approach be more helpful?

Yeah, I missed the one in `jmbase` which is now fixed, changes like that require the imports at the top to ensure `range` has the necessary python3 behavior(xrange is range in python3 but differs in python2 without those backport imports) so I wasn't changing it in modules until I was ready to add those imports and fix the encoding errors they cause. Right now I'm mostly trying to just reduce the amount of encoding issues and other obvious style issues.

jameshilliard commented:
> As for the files here, I guess you're addressing some but not all of the incompatibilities in specifically jmbase and jmdaemon, is that right?

For the moment yes, mostly since the next one I was looking into `jmbitcoin` is especially tricky.

> To echo what I've probably written elsewhere, when it comes to jmbitcoin at least, it probably makes sense to fold in the upgrade to Python3 compatibility with switching the binding to libsecp256k1 if possible, see coincurve (although such of thing cannot, of course, be done lightly).

It's usually a good idea to keep that sort of change separate so that you don't break a bunch of tests, it also means regressions can be more difficult to isolate since it's harder to be sure if a particular break is due to the python3 changes or due to the binding changes. Most of the encoding issues I was seeing in `jmbitcoin` were not issues with the binding but with the transaction parser encoding/decoding. This is mostly due to it relying on a lot of python2 `str` automatic conversion weirdness which needs to be made into explicit `bytes` or `unicode` conversions for python3.

> There is also an argument for a full-on rewrite of jmclient and jmdaemon using asyncio ... I don't really have the knowledge to assess this idea, but clearly it sounds like a lot of work.

Well for this project you would probably be best off using twisted's newer styles, but that would be something I would wait on until after dropping python2 support since we don't want to break the python2 test suite until the conversion is complete.

> This might also be a time to overhaul tests, which I made initially in a pretty messy way, tried to improve at one time, but then they suffered from pretty serious code-rot over the years.

I would try and do this as incrementally as possible, they do need to be tweaked for the universal python style somewhat but I'd be wary about doing a major overhaul of them at the same time as the conversion as opposed to trying to minimize the diff until that's complete.

> in case we accidentally end up putting efforts into porting code that isn't going to end up being used

It's probably easier to port as much as possible since you can probably validate more internal assumptions by having more test code, even if the features get dropped eventually.

jameshilliard commented:
> Changing all str occurrences in the Command.arguments and Command.response containers to bytes seems wrong to me. The twisted documentation doesn't explicitly state what data type those "names" should be, but "name", to me, implies a str, not bytes. On what basis did you change that?

Looking at this again I'm pretty sure bytes is actually required for python3(I recall running into that when trying to run it under python3), twisted made that [change](https://github.com/twisted/twisted/commit/e96142207eed981d61549daefe6c50ec7ee890b3#diff-838ce72be3738e4ac2b7366b8754362c) specifically for python3.

jameshilliard commented:
> Re: BigString, it seems that in the (nearly 2 years) time since I wrote this, the documentation is updated and it appears that AMP v2 supports arbitrary sized strings intrinsically anyway now.

I've gone ahead and removed `BigString()` and replaced it with `Unicode()`

jameshilliard commented:
> To echo what I've probably written elsewhere, when it comes to jmbitcoin at least, it probably makes sense to fold in the upgrade to Python3 compatibility with switching the binding to libsecp256k1 if possible, see coincurve (although such of thing cannot, of course, be done lightly). Since secp256k1-py is no longer maintained (which is more relevant than in some other cases, given the status of libsecp256k1).

Turns out coincurve was pretty much a drop in replacement for secp256k1-py, I've got all tests passing with it in #223. It should probably be reviewed and merged before we convert `jmbitcoin` to python3 style however.

undeath commented:
@jameshilliard: thank you for the clarifications regarding twisted. I guess they just forgot to update the documentation after porting to py3.

PR looks good to me.

jameshilliard commented:
Rebased with merge conflicts fixed.

mecampbellsoup commented:
Hm, having a slight challenge getting setup w/ PY3: ```
(jmvenv-py3) λ Matts-MBP scripts → λ git master* → python sendpayment.py -m 1 bitfinex.jmdat 4000000000 19difhWqqEa1pCcVTjK8Un3rvutzTk8zxd
Traceback (most recent call last):
  File "sendpayment.py", line 15, in <module>
    from jmclient import Taker, load_program_config, get_schedule,\
  File "/Users/mattysoup/Sites/Code/joinmarket-clientserver/jmclient/jmclient/__init__.py", line 9, in <module>
    from btc import *
ModuleNotFoundError: No module named 'btc'
```

Already remade the venv and ran `python setupall.py --develop`. What am I missing? Thanks :)

undeath commented:
You'd probably need to change that line to `from .btc import *`.

As was discussed with @AdamISZ on IRC before, `jmclient/btc.py` should be entirely removed. It contains lots of duplicated code from the jmbitcoin module.

mecampbellsoup commented:
I also hit this import error: https://github.com/yunabe/lgo/issues/42

https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/c08b69ba7558f705614510b22949827626aa92ff/jmbase/jmbase/__init__.py#L8

AdamISZ commented:
@mecampbellsoup i'm confused, are you trying to run this under Python 3? It's not python3 ready, it's a preliminary step.

mecampbellsoup commented:
Ah, okay, yea Python 3. I'll just checkout a release - was following master.

AdamISZ commented:
No, following master is fine if you want to do that, I'm saying, this is a preliminary set of code changes to help a transition to Py3, it's not ready to be actually run under Py3.

mecampbellsoup commented:
OK, well I got her initially because I pulled master and something broke with my PY2 setup, and upon reading the commits it looked like I would need to use Python 3. Will revert.

jameshilliard commented:
You may also need to rerun install.sh so that the `future` module gets installed.

AdamISZ commented:
Confirmed, I just updated to master and had the same exact thing :) If you don't want to install from scratch, just `pip install future` in your jmvenv should be enough (it was for me) @mecampbellsoup

jameshilliard commented:
You might also need to do `pip install txtorcon` since that was also a recently added dependency.

jameshilliard commented:
@mecampbellsoup by the way current master should be mostly functional under python3 now, although I think some stuff like the GUI may not work yet

AdamISZ commented:
@jameshilliard @mecampbellsoup worth mentioning we haven't yet updated docs, if you want to try Python3 now (I've been running it a while, more testers very welcome!) you have to indicate it in the installation script with `./install.sh --develop -p python3` or similar. Everything should function as before in Py3 except, as mentioned, the Qt (which I'm working on now). Cheers.
